src/server/pfs/fuse/server.go

```diff
@@ -1210,21 +1229,120 @@ func sanitizeInputAndGetAlias(datumInput *pps.Input, c *client.APIClient) (map[s
 		} else {
 			datumInputsToMounts[pfsInput] = []string{input.Pfs.Name}
 		}
-
 		return nil
 	}); err != nil {
+		return true, err
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.DatumInput = datumInput
+		mm.DatumInputsToMounts = datumInputsToMounts
+	}()
+	return useGlobFile, nil
+}
+
+func (mm *MountManager) GetNextXDatums(numDatums int, paginationMarker string) ([]*pps.DatumInfo, error) {
+	req := &pps.ListDatumRequest{
+		Input:            mm.DatumInput,
+		Number:           int64(numDatums),
+		PaginationMarker: paginationMarker,
+	}
+
+	ctx, cf := context.WithCancel(mm.Client.Ctx())
+	defer cf()
+	client, err := mm.Client.PpsAPIClient.ListDatum(ctx, req)
+	if err != nil {
 		return nil, err
 	}
+	var dis []*pps.DatumInfo
+	for {
+		di, err := client.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return dis, nil
+			}
+			return nil, err
+		}
+		dis = append(dis, di)
+	}
+}
 
-	return datumInputsToMounts, nil
+func (mm *MountManager) GetDatums(useGlobFile bool) (retErr error) {
+	defer func() {
+		retErr = grpcutil.ScrubGRPC(retErr)
+	}()
+	// TODO: Try creating datums using GlobFile for now because it's faster than ListDatum.
+	// Once ListDatum pagination is efficient, we can remove this. #ListDatumPagination
+	if useGlobFile {
+		return mm.CreateDatums()
+	}
+	// If input spec has Join or Group, fall back to ListDatum
+	datums, err := mm.GetNextXDatums(NumDatumsPerPage, "")
+	if err != nil {
+		return err
+	}
+	if len(datums) == 0 {
+		return errors.New("no datums to mount")
+	}
+	func() {
+		mm.mu.Lock()
+		defer mm.mu.Unlock()
+		mm.Datums = datums
+		mm.PaginationMarker = datums[len(datums)-1].Datum.Id
+		if len(datums) < NumDatumsPerPage || NumDatumsPerPage == 0 {
+			mm.AllDatumsReceived = true
+		}
+	}()
+	return nil
 }
 
-func getCopyOfMapping(datumInputsToMounts map[string][]string) map[string][]string {
-	datumInputsToMountsCopy := map[string][]string{}
-	for k, v := range datumInputsToMounts {
-		datumInputsToMountsCopy[k] = v
+func (mm *MountManager) CreateDatums() error {
+	datumsAtLevel := make(map[int][][]*pps.DatumInfo)
+	if err := visitInput(mm.DatumInput, 0, func(input *pps.Input, level int) error {
+		if input.Pfs != nil {
+			// Glob file to get datums associated with this PFS input
+			commit := client.NewCommit(input.Pfs.Project, input.Pfs.Repo, input.Pfs.Branch, "")
+			datums := []*pps.DatumInfo{}
+			if err := mm.Client.GlobFile(commit, input.Pfs.Glob, func(fi *pfs.FileInfo) error {
+				datums = append(datums, &pps.DatumInfo{Data: []*pfs.FileInfo{fi}})
```

Comment:
what happens if the glob contains 0 datums? can we add a test for this?

Reply:
If there are 0 datums, we return an error "no datums to mount" a little below. I'll add a test.

Reply:
should have been more clear; i meant part of the input that we are traversing contains 0 datums. for example if we visit some PfsInput as part of a larger input spec that contains 0, if i'm understanding correctly we are simply appending nothing in this case and returning. is my understanding correct? if so, it seems like the better thing to do might be to skip over it and continue finding more datums.

Reply:
If the PFS input generates 0 datums, we append an empty `datums` list and return `nil` (on line 1331), which continues the DFS. The only case where we stop the DFS early is if we return an error (on line 1312), which occurred in the call to `GlobFile()`.